### PR TITLE
Feat: update segmentation/zoning for caf.tem

### PR DESCRIPTION
**Pull request by isaac-tfn**
# Changes

Update segmentation/zoning behaviour to support forecasting + trip-end workflows.

- Change ZoningSystem.translate() to return ctk.translation.ZoneCorrespondence (instead of a raw DataFrame) and tighten spatial-translation metadata validation.
- Extend segmentation/DVector capabilities (e.g., slice aggregation, additional DVector helpers, partial IPF support for pd.Series targets).
- Update segment definitions/lookups/translations in YAML/CSV and apply various test formatting cleanups.

# Tasks

- [ ] Have unittests been added (testing individual functions and their edge cases)
- [ ] Has documentation (including user guide) been updated
- [ ] Have new dependencies been added (or changed) in relevant `requirements.txt`
- [ ] Code linting, tests and other workflows pass

### CODE REVIEW

File: src/caf/base/segments/norms_uc_od.yml [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  16: NHBONCA_Int
+  17: EBCA_Ext_FM
+  18: EBCA_Ext_TO
+  19: EBNCA_Ext
+  20: HBWCA_Ext_FM
+  21: HBWCA_Ext_TO
+  22: HBWNCA_Ext
+  23: OCA_Ext_FM
+  24: OCA_Ext_TO
+  25: ONCA_Ext
```
Comment (MattBuckley-TfN) [inline src/caf/base/segments/norms_uc_od.yml]: This is unnecessary if the aliases are the same as the values
```suggestion
```

File: src/caf/base/segments/m.yml [hunk 2]
```diff
@@ -3,8 +3,8 @@ name: m
 values:
   1: Walk
   2: Cycle
-  3: Car driver
-  4: Car passenger
+  3: Car driver + passenger
+  4: Van driver + passenger
```
Comment (MattBuckley-TfN) [inline src/caf/base/segments/m.yml:7]: Is this change fixing a problem with the previous version of mode or is the new mode segment different from the old one. Will this cause issues loading an older DVector with a new version of caf.base? Do we need a method for storing version of a segment to show segments which change?

File: src/caf/base/segmentation.py [hunk 3]
```diff
@@ -256,6 +260,29 @@ def remove(self, segment: str) -> "SegmentationSlice":
...
+        -------
+        SegmentationSlice containing the passed in segments.
+        """
+        params = {seg: val for seg, val in self.data.items() if seg in segments}
+        if isinstance(segments, list):
+            segments = [
+                segment.name if isinstance(segment, Segment) else segment
+                for segment in segments
+            ]
+        return SegmentationSlice(params, segments)
```
[the remark this answers is not in the document]
Comment (MattBuckley-TfN) [reply src/caf/base/segmentation.py]: This comment from copilot seems correct, I think if you move params to after the if statement when segments is either a list of string or a Segmentation it should work correctly.
```suggestion
        if isinstance(segments, list):
            segments = [
                segment.name if isinstance(segment, Segment) else segment
                for segment in segments
            ]
        params = {seg: val for seg, val in self.data.items() if seg in segments}
        return SegmentationSlice(params, segments)
```
Looks like `Segmentation.__contains__` can handle strings so it should work correctly.

File: src/caf/base/segmentation.py [hunk 4]
```diff
@@ -481,11 +524,11 @@ def has_time_period_segments(self) -> bool:
...
         cls,
-        source: Union[Path, pd.DataFrame],
+        source: pd.DataFrame,
         segmentation: Segmentation,
         escalate_warning: bool = False,
         cut_read: bool = False,
-    ) -> tuple[Segmentation, bool]:
+    ) -> tuple[Segmentation, bool, pd.DataFrame]:
         """
         Validate a segmentation from either a path to a csv, or a dataframe.
```
Comment (MattBuckley-TfN) [inline src/caf/base/segmentation.py:534]: Docstrings needs updating, in the first line description and the `source` parameter

File: src/caf/base/data_structures.py [hunk 5]
```diff
@@ -985,6 +1033,18 @@ def _generic_dunder(
                     "To multiply them, one must be translated "
                     "to match the other."
                 )
+        elif isinstance(other.zoning_system, Sequence):
```
Comment (MattBuckley-TfN) [inline src/caf/base/data_structures.py:1039]: This entire if statement has got quite complex, I'm not sure if there's time now but, it might be worth restructuring to simplify it and perhaps splitting it out into other methods / functions if needed.

File: src/caf/base/data_structures.py [hunk 6]
```diff
@@ -1234,18 +1327,18 @@ def composite_zoning(
...
-            validated_zoning.column_name,
-        }:
-            try:
-                trans_vector = trans_vector.set_index(
-                    [self.zoning_system.column_name, validated_zoning.column_name]
-                )
-            except Exception as exc:
-                raise ValueError("required zones not found in trans vector.") from exc
+        elif isinstance(trans_vector, pd.DataFrame):
+            trans_vector = ctk.translation.ZoneCorrespondence(
```
Comment (MattBuckley-TfN) [inline src/caf/base/data_structures.py:1334]: I'm surprised MyPy hasn't complained about redefining `trans_vector` as `ZoneCorrespondence` type, it might be worth setting it to a different variable anyway to make it more obvious when reading that `trans_vector` switches from a `DataFrame` to `ZoneCorrespondence` at this point.

File: src/caf/base/data_structures.py [hunk 7]
```diff
@@ -692,7 +718,15 @@ def translate_zoning(
                     new_zoning, weighting=weighting, cache_path=cache_path
                 )
         else:
-            trans_vector = target_zone.validate_translation_data(new_zoning, trans_vector)
+            trans_vector = target_zone.validate_translation_data(
+                new_zoning, trans_vector
+            )
+            trans_vector = ctk.translation.ZoneCorrespondence(
```
Comment (MattBuckley-TfN) [inline src/caf/base/data_structures.py]: Similar to [this comment](https://github.com/Transport-for-the-North/caf.base/pull/117#discussion_r3355140044) I'm surprised MyPy hasn't complained about the change of type, but I think I would probably have a new variable here anyway to make it clearer to a reader that the type changes at this point

File: src/caf/base/data_structures.py [hunk 8]
```diff
@@ -806,16 +843,22 @@ def split_by_agg_zoning(
             trans = self.zoning_system.translate(agg_zoning)
         else:
             trans = self.zoning_system.validate_translation_data(agg_zoning, trans)
+            trans = ctk.translation.ZoneCorrespondence(
```
Comment (MattBuckley-TfN) [inline src/caf/base/data_structures.py:849]: Similar to [this comment](https://github.com/Transport-for-the-North/caf.base/pull/117#discussion_r3355140044), I'd probably make the change of type into a new variable

File: src/caf/base/data_structures.py [hunk 9]
```diff
@@ -1555,9 +1665,18 @@ def combine_from_dic(
         new_data = new_data.reorder_levels(new_segmentation.naming_order).sort_index()
 
         return cls(
-            segmentation=new_segmentation, import_data=new_data, zoning_system=zoning_system
+            segmentation=new_segmentation,
+            import_data=new_data,
+            zoning_system=zoning_system,
         )
 
+    def reorder_levels(self, new_order: list[str]):
```
Comment (MattBuckley-TfN) [inline src/caf/base/data_structures.py]: Missing return type
```suggestion
    def reorder_levels(self, new_order: list[str]) -> DVector:
```

File: src/caf/base/data_structures.py [hunk 10]
```diff
@@ -1637,6 +1761,12 @@ def filter_segment_value(
             cut_read=self._cut_read,
         )
 
+    def filter_segment_values(self, filter: dict[str, list[int]]):
```
Comment (MattBuckley-TfN) [inline src/caf/base/data_structures.py]: Missing return type
```suggestion
    def filter_segment_values(self, filter: dict[str, list[int]]) -> DVector:
```

File: src/caf/base/data_structures.py [hunk 11]
```diff
@@ -1637,6 +1761,12 @@ def filter_segment_value(
...
+        out = self
+        for seg, values in filter.items():
+            out = out.filter_segment_value(seg, values)
+        return out
+
     def drop_by_segment_values(self, segment_name: str, segment_values: list[int]):
```
Comment (MattBuckley-TfN) [inline src/caf/base/data_structures.py]: Missing return type
```suggestion
    def drop_by_segment_values(self, segment_name: str, segment_values: list[int]) -> DVector:
```

File: src/caf/base/data_structures.py [hunk 12]
```diff
@@ -1674,7 +1806,12 @@ def fill(self, find_value: float | int, infill_value: float | int):
...
 
     def translate_segment(
-        self, from_seg, to_seg, reverse=False, drop_from=True, _bypass_validation: bool = False
+        self,
+        from_seg,
+        to_seg,
+        reverse=False,
+        drop_from=True,
+        _bypass_validation: bool = False,
     ):
```
Comment (MattBuckley-TfN) [inline src/caf/base/data_structures.py:1820]: Missing type annotations

File: src/caf/base/data_structures.py [hunk 13]
```diff
@@ -1969,7 +2151,7 @@ def ipf(
         DVector matched to targets, rmse achieved.
         """
         # check DVectors compatible
-        targets = self.validate_ipf_targets(targets, cache_path=zone_trans_cache)
+        # targets = self.validate_ipf_targets(targets, cache_path=zone_trans_cache)
         new_dvec = self.copy()
```
[the remark this answers is not in the document]
Comment (MattBuckley-TfN) [reply src/caf/base/data_structures.py:2160]: Is commenting this out done on purpose? If so can the line just be deleted instead?

File: src/caf/base/data_structures.py [hunk 14]
```diff
@@ -2127,6 +2324,10 @@ def concat_from_dir(
                 except:
```
Comment (MattBuckley-TfN) [inline src/caf/base/data_structures.py]: Should except something specific or at least `Exception` to avoid catching syntax errors or keyboard interrupts

File: src/caf/base/data_structures.py [hunk 15]
```diff
@@ -2535,6 +2750,26 @@ def balance_by_segments(
...
 
+    def balance_protect_subset(
+        self,
+        other: DVector,
+        target_zone: ZoningSystem,
+        protected_subset: dict[str, list[int]],
+    ):
+        protected = self.filter_segment_values(protected_subset).data.sum()
+        non_protected = self.data.sum() - protected
+        other_non_protected = other.data.sum() - protected
```
[the remark this answers is not in the document]
Comment (MattBuckley-TfN) [reply src/caf/base/data_structures.py:2766]: Is copilot correct here?

**Review by MattBuckley-TfN — CHANGES_REQUESTED** [review]
Copilot gave some good comments worth fixing and I've added some additional ones, mostly tidying things up. Also some additional things where I couldn't comment on files directly:

- [x] Please could you add some more detail to the PR title, so it's clearer when used in release notes.
- [ ] caf.toolkit version needs updating to >=0.13.0 to use `ZoneCorrespondence` class.
- [ ] I've [commented](https://github.com/Transport-for-the-North/caf.base/issues/112#issuecomment-4620449838) on #112 with some suggestions to include that issue within this PRs changes if you've got time.

**Comment by MattBuckley-TfN** [comment]
One thing I missed from the above comments is the `uni` segment includes some exclusions for `p_hb` and `p_nhb` which don't longer exist as segments. These should probably be removed from the segments to avoid issues, or maybe updated to be exclusions with the `direction` and `direction_od` segments.

File: src/caf/base/data_structures.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/base/data_structures.py:2766]: I am 99% sure it is not, can talk this section through to be sure?

**Review by MattBuckley-TfN — CHANGES_REQUESTED** [review]
There's a few comments which have been missed:
- [ ] In `DVector.remove_segment` segments should be a copy of the naming_order so that it doesn't edit the original, see [copilot comment](https://github.com/Transport-for-the-North/caf.base/pull/117#discussion_r3286976806)
- [ ] `Segmentation.validate_segmentation` docstring needs updating to remove reference to path being passed as source (see [my comment](https://github.com/Transport-for-the-North/caf.base/pull/117#discussion_r3354832923))
- [ ] This question about mode segment, see [my comment](https://github.com/Transport-for-the-North/caf.base/pull/117#discussion_r3354494247)
- [ ] `DVector.calc_rmse` error if `IpfTarget.data` is a Series see [copilot comment](https://github.com/Transport-for-the-North/caf.base/pull/117#discussion_r3286976954)

**Review by MattBuckley-TfN — APPROVED** [review]
LGTM
